File: ios/Utils/Factory/AdRequestFactory.swift
import Foundation
import YandexMobileAds

class AdRequestFactory {
    private let locationFactory: LocationFactory
    
    convenience init() {
        let locationFactory = LocationFactory()
        self.init(locationFactory: locationFactory)
    }
    
    init(locationFactory: LocationFactory) {
        self.locationFactory = locationFactory
    }
    
    func createFromDictionary(_ dict: [String: Any]) -> AdRequest? {
        let adRequest = AdRequest()
        let mutableRequest = adRequest.mutableRequest
        
        mutableRequest.location = locationFactory.createFromDictionary(dict[Constants.Key.location] as? [String: Any])
        mutableRequest.contextQuery = dict[Constants.Key.contextQuery] as? String
        mutableRequest.contextTags = dict[Constants.Key.contextTags] as? [String]
        if let biddingData = dict[Constants.Key.biddingData] as? String, !biddingData.isEmpty {
            mutableRequest.biddingData = biddingData
        }
        mutableRequest.parameters = dict[Constants.Key.parameters] as? [String: String]
        
        if let ageString = dict[Constants.Key.age] as? String,
           let ageInt = Int(ageString) {
            mutableRequest.age = NSNumber(value: ageInt)
        }
        
        if let gender = dict[Constants.Key.gender] as? String {
            switch gender {
            case Constants.Gender.male:
                mutableRequest.gender = kYMAGenderMale
            case Constants.Gender.female:
                mutableRequest.gender = kYMAGenderFemale
            default:
                mutableRequest.gender = nil
            }
        }
        
        mutableRequest.adTheme = .unspecified
        if let adTheme = dict[Constants.Key.adTheme] as? String {
            switch adTheme {
            case Constants.AdTheme.light:
                mutableRequest.adTheme = .light
            case Constants.AdTheme.dark:
                mutableRequest.adTheme = .dark
            default:
                break
            }
        }
        
        return mutableRequest
    }
}

File: android/src/main/java/com/yandexmobileads/util/factory/AdRequestFactory.kt
package com.yandexmobileads.util.factory

import com.facebook.react.bridge.ReadableMap
import com.yandex.mobile.ads.common.AdRequest
import com.yandex.mobile.ads.common.Gender
import com.yandex.mobile.ads.common.AdTheme
import com.yandexmobileads.common.Constants
import com.yandexmobileads.util.Converter

internal class AdRequestFactory(
    private val locationFactory: LocationFactory = LocationFactory(),
    private val converter: Converter = Converter(),
) {
    fun createFromReadableMap(readableMap: ReadableMap): AdRequest {
        val adRequest = AdRequest.Builder()

        adRequest.setAge(readableMap.getString(Constants.Key.AGE))
        adRequest.setLocation(
            locationFactory.createFromReadableMap(
                readableMap.getMap(Constants.Key.LOCATION)
            )
        )
        adRequest.setContextQuery(readableMap.getString(Constants.Key.CONTEXT_QUERY))
        adRequest.setContextTags(
            converter.convertReadableArrayToList(
                readableMap.getArray(
                    Constants.Key.CONTEXT_TAGS
                )
            )
        )
        adRequest.setGender(
            when (readableMap.getString(Constants.Key.GENDER)) {
                Constants.Gender.MALE -> Gender.MALE
                Constants.Gender.FEMALE -> Gender.FEMALE
                else -> null
            }
        )
        adRequest.setPreferredTheme(
            when (readableMap.getString(Constants.Key.AD_THEME)) {
                Constants.AdTheme.LIGHT -> AdTheme.LIGHT
                Constants.AdTheme.DARK -> AdTheme.DARK
                else -> null
            }
        )
        val biddingData = readableMap.getString(Constants.Key.BIDDING_DATA)
        if (biddingData.isNullOrEmpty().not()) {
            adRequest.setBiddingData(biddingData)
        }
        adRequest.setParameters(
            converter.convertReadableMapToMap(
                readableMap.getMap(Constants.Key.PARAMETERS)
            )
        )

        return adRequest.build()
    }
}
